crates/rattler/src/validation.rs

```diff
@@ -0,0 +1,298 @@
+    }
+
+    Ok(())
+}
+
+/// Determine whether the information in the [`PathsEntry`] matches the file in the package directory.
+fn validate_package_entry(
+    package_dir: &Path,
+    entry: &PathsEntry,
+) -> Result<(), PackageEntryValidationError> {
+    let path = package_dir.join(&entry.relative_path);
+
+    // Get the metadata for the entry
+    let metadata = match std::fs::symlink_metadata(&path) {
+        Ok(metadata) => metadata,
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+            return Err(PackageEntryValidationError::NotFound)
+        }
+        Err(e) => return Err(PackageEntryValidationError::GetMetadataFailed(e)),
+    };
+
+    // Validate based on the type of path
+    match entry.path_type {
+        PathType::HardLink => validate_package_hard_link_entry(path, entry, metadata),
+        PathType::SoftLink => validate_package_soft_link_entry(path, entry, metadata),
+        PathType::Directory => validate_package_directory_entry(path, entry, metadata),
+    }
+}
+
+/// Determine whether the information in the [`PathsEntry`] matches the file at the specified path.
+fn validate_package_hard_link_entry(
+    path: PathBuf,
+    entry: &PathsEntry,
+    metadata: Metadata,
+) -> Result<(), PackageEntryValidationError> {
+    debug_assert!(entry.path_type == PathType::HardLink);
+
+    // Validate the size of the file
+    if let Some(size_in_bytes) = entry.size_in_bytes {
+        if size_in_bytes != metadata.len() {
+            return Err(PackageEntryValidationError::IncorrectSize(
+                size_in_bytes,
+                metadata.len(),
+            ));
+        }
+    }
+
+    // Check the SHA256 hash of the file
+    if let Some(hash_str) = entry.sha256.as_deref() {
+        // Determine the hash of the file on disk
+        let hash = compute_file_sha256(&path)?;
+
+        // Convert the hash to bytes.
+        let mut expected_hash = <sha2::digest::Output<Sha256>>::default();
+        hex::decode_to_slice(hash_str, &mut expected_hash).map_err(|_| {
+            PackageEntryValidationError::HashMismatch(hash_str.to_owned(), format!("{:x}", hash))
+        })?;
+
+        // Compare the two hashes
+        if expected_hash != hash {
+            return Err(PackageEntryValidationError::HashMismatch(
+                hash_str.to_owned(),
+                format!("{:x}", hash),
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+/// Determine whether the information in the [`PathsEntry`] matches the symbolic link at the specified
+/// path.
+fn validate_package_soft_link_entry(
+    _path: PathBuf,
+    entry: &PathsEntry,
+    metadata: Metadata,
+) -> Result<(), PackageEntryValidationError> {
+    debug_assert!(entry.path_type == PathType::SoftLink);
+
+    if !metadata.is_symlink() {
+        return Err(PackageEntryValidationError::ExpectedSymlink);
+    }
+
+    // TODO: Validate symlink content
```

Comment:
I think in mamba we have a bit of an optimization to not recompute the SHA256 of a file that we already checked (e.g. if the symlink is going to a file of the same package).

Reply:
Yeah, I dont think we need to check the SHA256 hash of the file it points to because it *must* point to another file in the same package (which is then already checked). But we could also check whether the symlink itself is correct. E.g. `../a` points to the same file as `./../a` but the symlinks themselves are different.

I doubt it really matters though because during linking the symlinked files are symlinked in the target folder again anyway. But for completeness, I left the comment in.